### PR TITLE
Build functioning dev packages for IREECompiler and IREERuntime.

### PR DESCRIPTION
**Pull request by stellaraccident**
With this patch, external projects can do:

```
find_package(IREECompiler)
find_package(IREERuntime)
```

and access our targets by adding either the build or install tree `lib/cmake/IREE` to the CMake package search path.

There are still some ergonomic issues that need to be addressed going forward:

* cpuinfo breaks the one downstream I am testing this against. We really should just finish killing that, so I am setting -DIREE_ENABLE_CPUINFO=OFF in the dist package.
* We are not yet re-establishing the pretty `iree::runtime:...` aliases. The underscore versions are available (`iree_runtime_...`).
* I want to further isolate the remaining bundled deps (yaml and flatcc).
* New convenience install targets do not have full dependency information. Need to do an `all` build first.

As part of this, I did some longstanding cleanups:

* Normalized the install component namespace. Defined `IREECMakeExports`, `IREEDevLibraries-Compiler`, `IREEDevLibraries-Runtime`, `IREERuntimeLibraries-Compiler`, `IREEBundledLibraries`, `IREETools-Runtime`.
* Added convenience `iree-install-*` targets. Each includes a `-stripped` variant as well.
* Moved MLIRInterop.h to the bindings/c directory where it should have been all along and straightened the deps.
* Install the mlir-c headers to support libIREECompiler.so dev.
* Added the "busybox" tool binaries to the install distribution (iree-lld, etc).
* Stopped bundling upstream `lld` and `clang` by default because they bloat the size. `lld` is now bundled as `iree-lld` just like in the Python packages. Clang and llvm-link can be gotten by installing the new `IREETools-CompilerExtra` component if folks need those.

Note that this only installs public headers and libraries needed to build against the compiler and runtime. Installing private things needed to develop parts of the system are out of scope.

### CODE REVIEW

File: build_tools/third_party/tracy_client/CMakeLists.txt [hunk 1]
```diff
@@ -4,6 +4,13 @@
 # See https://llvm.org/LICENSE.txt for license information.
 # SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
 
+# Traced builds require the headers.
+set(IREE_HDRS_ROOT_PATH "${IREE_ROOT_DIR}/third_party/tracy/public")
```
Comment (benvanik) [inline build_tools/third_party/tracy_client/CMakeLists.txt:8]: we could fix this by duplicating some of the tracy C API - wouldn't be a lot of work and would let us hide this if you think it's worth it

Comment (stellaraccident) [reply build_tools/third_party/tracy_client/CMakeLists.txt:8]: Yeah, I had a similar thought but fine for the moment. I have more work to do on the bundled deps stuff and this patch is a big enough lift as it is.

**Review by benvanik — APPROVED** [review]

File: tools/iree-lld-main.cc [hunk 2]
```diff
@@ -0,0 +1,9 @@
+// Copyright 2021 The IREE Authors
```
Comment (rsuderman) [inline tools/iree-lld-main.cc:1]: Time travel!

**Review by rsuderman — APPROVED** [review]

File: tools/iree-lld-main.cc (hunk 2 above)
Comment (stellaraccident) [reply tools/iree-lld-main.cc:1]: (copied from that file)

**Pull request merged**